docs/RFCS/20180411_alter_column_type.md

```diff
@@ -0,0 +1,387 @@
+original, source, column while a backfill process is run. Once the
+backfill is complete, the shadow column is swapped for the original,
+which is then retired. The shadow column will be generally invisible to
+the user until it is swapped in. Shadow columns will be created in the
+same column family as the source column.
+
+This list describes the general type-change case, not all of these steps
+are necessarily executed for trivial or validated changes.
+
+* `ALTER ... COLUMN TYPE` is run, resulting in a new shadow column being
+  created. The shadow column will be created as a computed column, in
+  the source column's family, with the expression provided by the user
+  via `USING expression` or inferred for cast-type conversions.
+  * If there are indexes that refer to the column, additional
+    shadow-indexes are created.
+* The shadow column and indexes are backfilled using existing
+  functionality for creating
+  [new computed columns](https://github.com/cockroachdb/cockroach/issues/22652)
+  and indexes.
+* Nodes that write to the source column will synchronously apply the
+  requested type change and duplicate the write to the shadow column and
+  indexes.
+  * If the new value that is written is incompatible with the requested
+    conversion (e.g. `'foo'` -> `INET`), an error will be returned to
+    the caller.
+* Once the backfill is complete, several actions happen atomically:
+  * The computed nature is removed from the shadow column.
+  * The hidden flag is removed from the shadow column.
+  * The source column is dropped.
+  * The shadow column is renamed.
+  * Any indexes on the original column are dropped.
+  * Shadow indexes are renamed to replace the original indexes.
+
+## Failure modes
+
+Changing a column type is a low-risk activity, at least from the
+perspective of data safety, because the process can be aborted at any
+time up until the shadow column and indexes are made live. All
+validation or conversion of existing will have been performed before the
+new type becomes visible to SQL consumers, while any writes performed
+during the conversion process must conform to both the old and new
+column types.
+
+The most likely cause for type-change failure is the presence of
+existing data which cannot be coerced into the desired type. In this
+case, the type change will abort and a sample of non-conforming rows /
+values should be returned to the user.
+
+Type changes which require extensive rewriting of data will temporarily
+increase the amount of disk space required for the new sstables until
+the old data can be retired. The amount of extra disk space required
+could be estimated from existing table stats.
+
+In the case where a column-type conversion (e.g. `STRING -> UUID`) is in
+process and a user attempts to insert data that is not compatible with
+the target type, the `INSERT` should fail with an error message such as
+`column user_id is being converted from STRING to UUID. The value
```

Comment:
if you reuse the computed column mechanism, it may not be practical / elegant to change the error message. Please check what would currently happen with a computed column, and see if the error message is acceptable from a UX perspective. If it's not, then this RFC should outline how you plan to extend the code for error handling, and/or if it would be possible to use the same error messages for both regular computed columns and type substitutions as described here. I would prefer to avoid weird boolean flags throughout multiple layers of abstraction just to get the right error message.